### ISSUE TITLE
Upgrade from 4.2.4 to 4.3.0 gives EISDIR

### ISSUE THREAD

**Issue by frenkel** [thoughtbot/paperclip#1958]
In an app of ours we're duplicating photo attachments to a new database row using code like this:

```ruby
photo = @original.photos.first
newphoto = photo.dup
newphoto.file = photo.file
@other.photos << newphoto
newphoto.save
```

This used to work fine, but after upgrading to Paperclip 4.3.0 it broke with the error below in all cases (i.e. `@other.photos` is empty or already has some photos).

```
Errno::EISDIR in DiffController#update
Is a directory @ unlink_internal - <projectdir>/public

<rubydir>/fileutils.rb:1450:in `unlink'
<rubydir>/fileutils.rb:1450:in `block in remove_file'
<rubydir>/fileutils.rb:1455:in `platform_support'
<rubydir>/fileutils.rb:1449:in `remove_file'
<rubydir>/fileutils.rb:793:in `remove_file'
<rubydir>/fileutils.rb:571:in `block in rm'
<rubydir>/fileutils.rb:570:in `each'
<rubydir>/fileutils.rb:570:in `rm'
paperclip (4.3.0) lib/paperclip/storage/filesystem.rb:64:in `block in flush_deletes'
paperclip (4.3.0) lib/paperclip/storage/filesystem.rb:61:in `each'
paperclip (4.3.0) lib/paperclip/storage/filesystem.rb:61:in `flush_deletes'
paperclip (4.3.0) lib/paperclip/attachment.rb:239:in `save'
paperclip (4.3.0) lib/paperclip/has_attached_file.rb:92:in `block in add_active_record_callbacks'
activesupport (4.2.3) lib/active_support/callbacks.rb:444:in `instance_exec'
activesupport (4.2.3) lib/active_support/callbacks.rb:444:in `block in make_lambda'
activesupport (4.2.3) lib/active_support/callbacks.rb:226:in `call'
activesupport (4.2.3) lib/active_support/callbacks.rb:226:in `block in halting_and_conditional'
activesupport (4.2.3) lib/active_support/callbacks.rb:504:in `call'
activesupport (4.2.3) lib/active_support/callbacks.rb:504:in `block in call'
activesupport (4.2.3) lib/active_support/callbacks.rb:504:in `each'
activesupport (4.2.3) lib/active_support/callbacks.rb:504:in `call'
activesupport (4.2.3) lib/active_support/callbacks.rb:88:in `run_callbacks'
activerecord (4.2.3) lib/active_record/callbacks.rb:302:in `create_or_update'
activerecord (4.2.3) lib/active_record/persistence.rb:120:in `save'
activerecord (4.2.3) lib/active_record/validations.rb:37:in `save'
activerecord (4.2.3) lib/active_record/attribute_methods/dirty.rb:21:in `save'
activerecord (4.2.3) lib/active_record/transactions.rb:286:in `block (2 levels) in save'
activerecord (4.2.3) lib/active_record/transactions.rb:351:in `block in with_transaction_returning_status'
activerecord (4.2.3) lib/active_record/connection_adapters/abstract/database_statements.rb:211:in `transaction'
activerecord (4.2.3) lib/active_record/transactions.rb:220:in `transaction'
activerecord (4.2.3) lib/active_record/transactions.rb:348:in `with_transaction_returning_status'
activerecord (4.2.3) lib/active_record/transactions.rb:286:in `block in save'
activerecord (4.2.3) lib/active_record/transactions.rb:301:in `rollback_active_record_state!'
activerecord (4.2.3) lib/active_record/transactions.rb:285:in `save'
activerecord (4.2.3) lib/active_record/associations/has_many_association.rb:41:in `insert_record'
activerecord (4.2.3) lib/active_record/associations/collection_association.rb:569:in `block (2 levels) in concat_records'
activerecord (4.2.3) lib/active_record/associations/collection_association.rb:408:in `replace_on_target'
activerecord (4.2.3) lib/active_record/associations/collection_association.rb:403:in `add_to_target'
activerecord (4.2.3) lib/active_record/associations/collection_association.rb:568:in `block in concat_records'
activerecord (4.2.3) lib/active_record/associations/collection_association.rb:566:in `each'
activerecord (4.2.3) lib/active_record/associations/collection_association.rb:566:in `concat_records'
activerecord (4.2.3) lib/active_record/associations/has_many_association.rb:173:in `concat_records'
activerecord (4.2.3) lib/active_record/associations/collection_association.rb:168:in `block in concat'
activerecord (4.2.3) lib/active_record/associations/collection_association.rb:183:in `block in transaction'
activerecord (4.2.3) lib/active_record/connection_adapters/abstract/database_statements.rb:213:in `block in transaction'
activerecord (4.2.3) lib/active_record/connection_adapters/abstract/transaction.rb:184:in `within_new_transaction'
activerecord (4.2.3) lib/active_record/connection_adapters/abstract/database_statements.rb:213:in `transaction'
activerecord (4.2.3) lib/active_record/transactions.rb:220:in `transaction'
activerecord (4.2.3) lib/active_record/associations/collection_association.rb:182:in `transaction'
activerecord (4.2.3) lib/active_record/associations/collection_association.rb:168:in `concat'
activerecord (4.2.3) lib/active_record/associations/collection_proxy.rb:970:in `<<'
app/controllers/diff_controller.rb:266:in `block in update'
activerecord (4.2.3) lib/active_record/relation/delegation.rb:46:in `each'
activerecord (4.2.3) lib/active_record/relation/delegation.rb:46:in `each'
app/controllers/diff_controller.rb:251:in `update'
```

I duplicating not supposed to be done like this? Or did something break?

**Comment by tute** [comment]
A minor version bump shouldn't break any existing code, so I consider that a bug. @jyurek do you know what could be happening here?

**Comment by tute** [comment]
@frenkel can you spot the commit that introduced this regression, so we revert or fix it? Thanks!

**Comment by frenkel** [comment]
Using this timeline: https://github.com/thoughtbot/paperclip/compare/v4.2.2...v4.3.0

I found that commit 9a6919404a1e7d61d98aab0749fc25847ed0244a is the first one that fails.

**Comment by tute** [comment]
We released 4.3.1, that includes a fix: 36095bc0701a65054726af55b2689df9f7e75b9c. Can you do the patch upgrade of paperclip, and see if it fixes this issue? Thank you!

**Comment by frenkel** [comment]
Yes, this fixes the problem. Thanks!
